stl/inc/__msvc_int128.hpp

```diff
@@ -0,0 +1,1155 @@
+        uint32_t (&__q)[__m_plus_n_plus_one - __n]) noexcept {
+        _Knuth_4_3_1_D_impl(__u, __v, __q, __n, __m_plus_n_plus_one - (__n + 1));
+    }
+
+    _NODISCARD static constexpr uint64_t _UDiv128(
+        uint64_t _High, uint64_t _Low, uint64_t _Div, uint64_t& _Remainder) noexcept {
+        _STL_INTERNAL_CHECK(_High < _Div);
+
+#if !defined(__clang__) && defined(_M_X64) && !defined(_M_ARM64EC)
+        if (!_STD is_constant_evaluated()) {
+            return _udiv128(_High, _Low, _Div, &_Remainder);
+        }
+#endif // !defined(__clang__) && defined(_M_X64) && !defined(_M_ARM64EC)
+
+        const auto __d = _STD countl_zero(static_cast<uint32_t>(_Div >> 32));
+        if (__d >= 32) { // _Div < 2^32
+            auto _Rem    = (_High << 32) | (_Low >> 32);
+            auto _Result = _Rem / static_cast<uint32_t>(_Div);
+            _Rem         = (_Rem % static_cast<uint32_t>(_Div) << 32) | static_cast<uint32_t>(_Low);
+            _Result      = _Result << 32 | _Rem / static_cast<uint32_t>(_Div);
+            _Remainder   = _Rem % static_cast<uint32_t>(_Div);
+            return _Result;
+        }
+
+        uint32_t __u[5] = {
+            static_cast<uint32_t>(_Low << __d),
+            static_cast<uint32_t>(_Low >> (32 - __d)),
+            static_cast<uint32_t>(_High << __d),
+            static_cast<uint32_t>(_High >> (32 - __d)),
+            0,
+        };
+        if (__d != 0) {
+            __u[2] |= static_cast<uint32_t>(_Low >> (64 - __d));
+            __u[4] = static_cast<uint32_t>(_High >> (64 - __d));
+        }
+        uint32_t __v[2] = {
+            static_cast<uint32_t>(_Div << __d),
+            static_cast<uint32_t>(_Div >> (32 - __d)),
+        };
+        uint32_t __q[3];
+
+        _Knuth_4_3_1_D(__u, __v, __q);
+        _STL_INTERNAL_CHECK(__u[4] == 0);
+        _STL_INTERNAL_CHECK(__u[3] == 0);
+        _STL_INTERNAL_CHECK(__u[2] == 0);
+        _Remainder = (static_cast<uint64_t>(__u[1]) << (32 - __d)) | (__u[0] >> __d);
+
+        _STL_INTERNAL_CHECK(__q[2] == 0);
+        return static_cast<uint64_t>(__q[1]) << 32 | __q[0];
+    }
+
+    constexpr _Uint128() noexcept : _Word{} {}
+
+    constexpr _Uint128(const integral auto _Val) noexcept : _Word{static_cast<uint64_t>(_Val)} {
+        if constexpr (signed_integral<decltype(_Val)>) {
+            if (_Val < 0) {
+                _Word[1] = ~0ull;
+            }
+        }
+    }
+
+    explicit constexpr _Uint128(const uint64_t _Low, const uint64_t _High) noexcept : _Word{_Low, _High} {}
+
+    template <integral _Ty>
+    _NODISCARD explicit constexpr operator _Ty() const noexcept {
+        return static_cast<_Ty>(_Word[0]);
+    }
+
+    _NODISCARD explicit constexpr operator bool() const noexcept {
+        return (_Word[0] | _Word[1]) != 0;
+    }
+
+    _NODISCARD constexpr friend bool operator==(const _Uint128& _Left, const _Uint128& _Right) noexcept {

```

Comment:
Would the `_NODISCARD_FRIEND` macro be appropriate here (and several places below)?

If not, there's also some inconsistency between `_NODISCARD constexpr friend` here and `_NODISCARD friend constexpr` below.

Reply:
> Would the `_NODISCARD_FRIEND` macro be appropriate here (and several places below)?

Sure, that seems like a good consistency change to me even if not strictly needed.

> If not, there's also some inconsistency between `_NODISCARD constexpr friend` here and `_NODISCARD friend constexpr` below.

Ouch.

> Given that this is a C++20 header, why not default the equality?

I went through a "union of 2 64-bit words and 4 32-bit words" phase, and it hadn't occurred to me to restore the defaulted equality when I returned to non-union land.

Reply:
Eww, and `explicit constexpr` got in here as well.